ktor-server/ktor-server-core/jvm/src/io/ktor/application/newapi/KtorPlugin.kt

```diff
@@ -0,0 +1,655 @@
+     * */
+    public fun finish(): Unit = context.finish()
+
+    /**
+     * A subject of the current execution. A subject can be any object of some allowed type that is currently being
+     * processed.
+     *
+     * - For the stage of processing a call, there is no [subject] (i.e. [SubjectT] is [Unit])
+     * - For the stage of sending a response, the [subject] is an object that is currently being sent via HTTP protocol. ([SubjectT] can be [Any] type)
+     * - For the stage of receiving a request data, the [subject] is of type [ApplicationReceiveRequest]. Please, read [ApplicationReceiveRequest] javadoc for more information.
+     * */
+    public val subject: SubjectT get() = context.subject
+
+    /**
+     * A call that is being processed by the plugin in the current execution.
+     * */
+    public val call: ApplicationCall get() = context.call
+
+    // Useful methods
+
+    /**
+     * See [ApplicationEnvironment] for more details.
+     * */
+    public val environment: ApplicationEnvironment get() = context.application.environment
+
+    /**
+     * Configuration of the curreny application.
+     * */
+    public val configuration: ApplicationConfig get() = environment.config
+
+    /**
+     * Port of the current application. Same as in config.
+     * */
+    public val port: Int get() = configuration.propertyOrNull("ktor.deployment.port")?.getString()?.toInt() ?: 8080
+
+    /**
+     * Host of the current application. Same as in config.
+     * */
+    public val host: String get() = configuration.propertyOrNull("ktor.deployment.host")?.getString() ?: "0.0.0.0"
+
+    /**
+     * Sets a shutdown hook. This method is useful for closing resources allocated by the feature.
+     * */
+    public fun onShutdown(block: suspend () -> Unit) {
+        GlobalScope.launch(context.coroutineContext) {
+            block()
+        }
+    }
+}
+
+/**
+ * [Execution] for the stage of processing a call.
+ * */
+public typealias CallExecution = Execution<Unit>
+
+/**
+ * [Execution] for the stage of receiving data in a call.
+ * */
+public typealias ReceiveExecution = Execution<ApplicationReceiveRequest>
+
+/**
+ * [Execution] for the stage of sending response to a call.
+ * */
+public typealias ResponseExecution = Execution<Any>
+
+/**
+ * A plugin for Ktor that embeds into the HTTP pipeline and extends functionality of Ktor framework.
+ * */
+public abstract class KtorPlugin<Configuration : Any> private constructor(
+    public override val name: String
+) : ApplicationFeature<ApplicationCallPipeline, Configuration, KtorPlugin<Configuration>>,
+    PluginContext,
+    InterceptionsHolder {
+
+    protected var configurationValue: Configuration? = null
+    protected var pipelineValue: ApplicationCallPipeline? = null
+
+    protected val pipeline: ApplicationCallPipeline get() = pipelineValue!!
+
+    public val plugin: Configuration get() = configurationValue!!
```

Comment:
Why is it named `plugin`? Should it be `config`?

Reply:
Can rename, no problem